File: dotnet/tests/PIIMasker.Tests/ModelStoreTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using Xunit;

namespace PIIMasker.Tests;

/// <summary>
/// Where the model is looked for, and what the loader last concluded about it.
/// </summary>
/// <remarks>
/// <para>
/// The counterpart of the Swift target's <c>ModelStoreTests</c>, plus the cases that only exist
/// here: this target accepts a <c>current</c> plain-text pointer as well as a symlink, and that
/// file is something a caller can write — so what it is allowed to name is a security boundary, not
/// a convenience.
/// </para>
/// <para>
/// The store is process-wide state, so these run in one collection and each test configures it.
/// </para>
/// </remarks>
[Collection(nameof(ModelStoreTests))]
[CollectionDefinition(nameof(ModelStoreTests), DisableParallelization = true)]
public sealed class ModelStoreTests : IDisposable
{
    private readonly string root = Path.Combine(
        Path.GetTempPath(), "piimasker-store-" + Guid.NewGuid().ToString("N")[..12]);

    public ModelStoreTests() => Directory.CreateDirectory(root);

    public void Dispose()
    {
        ModelStore.Configure(ModelStore.Location.Unconfigured);
        try { Directory.Delete(root, recursive: true); } catch (IOException) { } catch (UnauthorizedAccessException) { }
    }

    /// <summary>A directory that looks like a model to <see cref="ModelInstaller.IsCompleteModelDir"/>.</summary>
    private string CompleteDir(string name)
    {
        var directory = Path.Combine(root, name);
        Directory.CreateDirectory(directory);
        foreach (var file in ModelInstaller.RequiredModelFiles)
            File.WriteAllText(Path.Combine(directory, file), "x");
        return directory;
    }

    private void Promote(string version) =>
        File.WriteAllText(Path.Combine(root, "current"), version);

    /// <summary>
    /// A host that forgot to configure gets "no model" rather than a guess, so its fail-closed gate
    /// holds everything.
    /// </summary>
    [Fact]
    public void UnconfiguredResolvesNothing()
    {
        ModelStore.Configure(ModelStore.Location.Unconfigured);

        Assert.Null(ModelStore.ResolvedModelDirectory());
        Assert.False(ModelStore.ModelIsInstalled());
    }

    /// <summary>The install root wins, and it resolves through the promotion pointer.</summary>
    [Fact]
    public void TheInstallRootResolvesThroughThePromotionPointer()
    {
        CompleteDir("2026.08.1");
        Promote("2026.08.1");
        ModelStore.Configure(root);

        Assert.Equal(Path.Combine(root, "2026.08.1"), ModelStore.ResolvedModelDirectory());
        Assert.True(ModelStore.ModelIsInstalled());
    }

    /// <summary>
    /// Fallbacks are searched after the install root, in order, and each must be complete on its own.
    /// </summary>
    /// <remarks>
    /// Answering per file instead would let a directory missing one file still report installed by
    /// resolving that file from a different candidate — and the loader, which reads all three from
    /// one folder, would then fail with nothing to retry it.
    /// </remarks>
    [Fact]
    public void FallbacksAreOrderedAndMustEachBeComplete()
    {
        var incomplete = Path.Combine(root, "incomplete");
        Directory.CreateDirectory(incomplete);
        File.WriteAllText(Path.Combine(incomplete, ModelInstaller.RequiredModelFiles[0]), "x");
        var good = CompleteDir("shipped");
        var alsoGood = CompleteDir("checkout");

        ModelStore.Configure(new ModelStore.Location(null, [incomplete, good, alsoGood]));

        Assert.Equal(good, ModelStore.ResolvedModelDirectory());
    }

    /// <summary>
    /// A promotion pointing at an incomplete directory falls through to the fallbacks rather than
    /// resolving to something the loader cannot open.
    /// </summary>
    [Fact]
    public void AnIncompletePromotionFallsThrough()
    {
        Directory.CreateDirectory(Path.Combine(root, "2026.08.1"));
        Promote("2026.08.1");
        var shipped = CompleteDir("shipped");

        ModelStore.Configure(new ModelStore.Location(root, [shipped]));

        Assert.Equal(shipped, ModelStore.ResolvedModelDirectory());
    }

    /// <summary>Re-configuring throws away everything concluded about the previous location.</summary>
    [Fact]
    public void ReconfiguringInvalidates()
    {
        CompleteDir("2026.08.1");
        Promote("2026.08.1");
        ModelStore.Configure(root);
        Assert.True(ModelStore.ModelIsInstalled());
        ModelStore.MarkLoaded();
        Assert.Equal(ModelStore.State.Loaded, ModelStore.CurrentState);

        ModelStore.Configure(ModelStore.Location.Unconfigured);

        // The cached "installed" verdict went with it, or a relocated host would keep reporting a
        // model it can no longer find.
        Assert.False(ModelStore.ModelIsInstalled());
        Assert.Equal(ModelStore.State.Unknown, ModelStore.CurrentState);
    }

    /// <summary>A real change raises the event once; setting the same state again raises nothing.</summary>
    [Fact]
    public void StateChangesArePushedOnce()
    {
        ModelStore.Configure(ModelStore.Location.Unconfigured);
        var raised = 0;
        void Handler(object? sender, EventArgs e) => raised++;
        ModelStore.StateChanged += Handler;
        try
        {
            ModelStore.MarkLoaded();
            ModelStore.MarkLoaded();
            Assert.Equal(1, raised);

            ModelStore.MarkUnusable();
            Assert.Equal(2, raised);
        }
        finally { ModelStore.StateChanged -= Handler; }
    }

    /// <summary>
    /// Only the loader may certify. <see cref="ModelStore.Invalidate"/> returns the state to
    /// "nothing known", never to "fine".
    /// </summary>
    [Fact]
    public void InvalidatingReturnsToUnknownRatherThanHealthy()
    {
        ModelStore.MarkUnusable();
        Assert.Equal(ModelStore.State.Unusable, ModelStore.CurrentState);

        ModelStore.Invalidate();

        Assert.Equal(ModelStore.State.Unknown, ModelStore.CurrentState);
    }

    /// <summary>Every invalidation advances the generation, so a cached masker can tell it is stale.</summary>
    [Fact]
    public void InvalidationAdvancesTheGeneration()
    {
        var before = ModelStore.Generation;

        ModelStore.Invalidate();

        Assert.True(ModelStore.Generation > before,
            $"generation did not advance: {before} → {ModelStore.Generation}");
    }

    /// <summary>
    /// The promotion pointer may only name a plain directory entry inside the install root.
    /// </summary>
    /// <remarks>
    /// No Swift counterpart — Swift promotes with a symlink and has no file to validate. Here the
    /// file is on disk, so a caller able to write it must not be able to aim the loader, or a later
    /// recursive delete, at anything outside the root. Every one of these resolves to null rather
    /// than to a path.
    /// </remarks>
    [Theory]
    [InlineData("..")]
    [InlineData("../elsewhere")]
    [InlineData("..\\elsewhere")]
    [InlineData("/etc")]
    [InlineData("C:\\Windows")]
    [InlineData("sub/dir")]
    [InlineData("")]
    [InlineData("   ")]
    [InlineData("NUL")]
    [InlineData("COM1.2026")]
    public void APromotionPointerCannotAimOutsideTheInstallRoot(string named)
    {
        Promote(named);

        Assert.Null(ModelStore.ResolvePromoted(root));
    }

    /// <summary>A symlink-shaped promotion — a real <c>current</c> directory — resolves too.</summary>
    /// <remarks>
    /// That is the Swift target's layout, and a store written by either target has to resolve in
    /// both, or a user moving between them silently loses their installed model.
    /// </remarks>
    [Fact]
    public void ADirectoryNamedCurrentResolvesAsTheSwiftLayout()
    {
        var current = Path.Combine(root, "current");
        Directory.CreateDirectory(current);

        Assert.Equal(current, ModelStore.ResolvePromoted(root));
    }

    /// <summary>Nothing promoted is not an error; it is "no model yet".</summary>
    [Fact]
    public void NoPromotionResolvesToNull()
    {
        Assert.Null(ModelStore.ResolvePromoted(root));
    }

    /// <summary>Surrounding whitespace in the pointer is not a different version.</summary>
    [Fact]
    public void ThePromotionPointerIsTrimmed()
    {
        CompleteDir("2026.08.1");
        Promote("  2026.08.1\r\n");

        Assert.Equal(Path.Combine(root, "2026.08.1"), ModelStore.ResolvePromoted(root));
    }
}

File: swift/Tests/PIIMaskerTests/ModelStoreTests.swift
import Foundation
import Testing

@testable import PIIMasker

/// `ModelStore` is process-wide by design, so these run serialized — a parallel test reconfiguring
/// the location underneath another would make both flaky for reasons that have nothing to do with
/// the code.
///
/// No weights needed: every case here is about which directory is chosen and what the loader's
/// verdict is, not about inference.
@Suite("Model store", .serialized)
struct ModelStoreTests {

    /// Restores whatever the store was configured with, so one test can't leak into the next.
    static func withLocation(_ location: ModelStore.Location, _ body: () throws -> Void) rethrows {
        let saved = ModelStore.location
        defer { ModelStore.configure(saved) }
        ModelStore.configure(location)
        try body()
    }

    static func completeDir(_ url: URL, marker: String = "weights") throws -> URL {
        try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        try marker.data(using: .utf8)!.write(to: url.appendingPathComponent("model.onnx"))
        for f in ModelInstaller.requiredModelFiles where f != "model.onnx" {
            try "{}".data(using: .utf8)!.write(to: url.appendingPathComponent(f))
        }
        return url
    }

    /// A host that never configured the store resolves nothing — so the fail-closed gate holds
    /// everything. Safe, and it looks exactly like a missing download, which is the correct story.
    @Test func unconfiguredResolvesNothing() {
        Self.withLocation(.unconfigured) {
            #expect(ModelStore.resolvedModelDirectory() == nil)
            #expect(!ModelStore.modelIsInstalled())
        }
    }

    /// The install root's `current` wins over every fallback, and it is resolved THROUGH the symlink
    /// so a caller holds a concrete version directory: a refresh flipping `current` afterwards
    /// cannot change what an in-flight load is reading.
    @Test func installRootWinsAndResolvesThroughTheSymlink() throws {
        let root = TestPublish.scratch(); let trash = Trash(root)
        let install = root.appendingPathComponent("install")
        let bundled = root.appendingPathComponent("bundled")
        _ = try Self.completeDir(install.appendingPathComponent("v1"), marker: "installed")
        _ = try Self.completeDir(bundled, marker: "bundled")
        try ModelInstaller.pointCurrent(at: "v1", in: install)

        Self.withLocation(ModelStore.Location(installRoot: install, fallbacks: [bundled])) {
            let dir = ModelStore.resolvedModelDirectory()
            #expect(dir?.lastPathComponent == "v1", "\(dir?.path ?? "nil")")
            let weights = (try? String(
                contentsOf: dir!.appendingPathComponent("model.onnx"), encoding: .utf8)) ?? ""
            #expect(weights == "installed", "\(weights)")
        }
        _ = trash
    }

    /// Fallbacks are searched in order, and only a COMPLETE directory counts. Answering per file
    /// would let a directory missing one of them still report installed by resolving that file from
    /// a different candidate — and the loader, which reads them from a single folder, would then
    /// fail with nothing to retry it.
    @Test func fallbacksAreOrderedAndMustBeComplete() throws {
        let root = TestPublish.scratch(); let trash = Trash(root)
        let incomplete = root.appendingPathComponent("incomplete")
        try FileManager.default.createDirectory(at: incomplete, withIntermediateDirectories: true)
        try "x".data(using: .utf8)!.write(to: incomplete.appendingPathComponent("model.onnx"))
        let good = try Self.completeDir(root.appendingPathComponent("good"), marker: "good")

        Self.withLocation(
            ModelStore.Location(installRoot: nil, fallbacks: [incomplete, good])
        ) {
            #expect(ModelStore.resolvedModelDirectory()?.lastPathComponent == "good")
            #expect(ModelStore.modelIsInstalled())
        }
        _ = trash
    }

    /// An install root whose `current` is incomplete falls through to the fallbacks rather than
    /// reporting a model that cannot load.
    @Test func anIncompleteCurrentFallsThrough() throws {
        let root = TestPublish.scratch(); let trash = Trash(root)
        let install = root.appendingPathComponent("install")
        let broken = install.appendingPathComponent("v1")
        try FileManager.default.createDirectory(at: broken, withIntermediateDirectories: true)
        try "x".data(using: .utf8)!.write(to: broken.appendingPathComponent("model.onnx"))
        try ModelInstaller.pointCurrent(at: "v1", in: install)
        let bundled = try Self.completeDir(root.appendingPathComponent("bundled"), marker: "bundled")

        Self.withLocation(ModelStore.Location(installRoot: install, fallbacks: [bundled])) {
            #expect(ModelStore.resolvedModelDirectory()?.lastPathComponent == "bundled")
        }
        _ = trash
    }

    /// Reconfiguring invalidates: the bytes the loader last judged are not the bytes at the new
    /// location, so a cached "installed" answer from the old one must not carry over.
    @Test func reconfiguringInvalidates() throws {
        let root = TestPublish.scratch(); let trash = Trash(root)
        let good = try Self.completeDir(root.appendingPathComponent("good"))

        Self.withLocation(ModelStore.Location(installRoot: nil, fallbacks: [good])) {
            #expect(ModelStore.modelIsInstalled())
            ModelStore.configure(.unconfigured)
            #expect(!ModelStore.modelIsInstalled(), "the cached positive must not survive")
        }
        _ = trash
    }

    /// A load failure has to reach the host the moment it happens. The load runs asynchronously off
    /// launch, so whether it beats the host's last badge refresh is a race — and it loses whenever
    /// the ONNX session builds and only the tokenizer fails. Without this push the user sees a
    /// healthy-looking app with no badge and no retry.
    @Test func stateChangesArePushedOnce() {
        let saved = ModelStore.onStateChange
        defer { ModelStore.onStateChange = saved; ModelStore.invalidate() }

        ModelStore.invalidate()
        let pushes = Flag()
        ModelStore.onStateChange = { pushes.value = true }

        ModelStore.markUnusable()
        #expect(pushes.value)

        pushes.value = false
        ModelStore.markUnusable()
        #expect(!pushes.value, "an unchanged state should push nothing")

        ModelStore.invalidate()
        #expect(pushes.value, "clearing it should push too")
    }

    /// `markUnusable` also drops the installed cache: the files are there, so a cached positive
    /// would keep reporting a healthy install for a model nothing can open.
    @Test func markUnusableClearsTheInstalledCache() throws {
        let root = TestPublish.scratch(); let trash = Trash(root)
        let good = try Self.completeDir(root.appendingPathComponent("good"))

        Self.withLocation(ModelStore.Location(installRoot: nil, fallbacks: [good])) {
            #expect(ModelStore.modelIsInstalled())
            ModelStore.markUnusable()
            #expect(ModelStore.state == .unusable)
            // The files are still there, so it re-checks disk and answers true again — but it did
            // re-check, which is the point.
            #expect(ModelStore.modelIsInstalled())
        }
        ModelStore.invalidate()
        _ = trash
    }

    // MARK: - The fail-closed gate

    /// With no model, `maskFields` returns nil rather than the values it was handed. This is the
    /// whole contract: the caller's only correct response to nil is to send nothing.
    @Test func maskFieldsHoldsWhenNoModelIsInstalled() async {
        let saved = ModelStore.location
        defer { ModelStore.configure(saved) }
        ModelStore.configure(.unconfigured)

        let masker = PrivacyFilter()
        let out = await masker.maskFields(["a@b.com", "call 415-555-0142"], timeout: 1)
        #expect(out == nil)
    }

    /// Nothing to mask is success, not a drop — a caller with no populated fields must not be told
    /// to abandon the request.
    @Test func noValuesIsSuccess() async {
        let masker = PrivacyFilter()
        let out = await masker.maskFields([], timeout: 1)
        #expect(out?.masked.isEmpty == true)
        #expect(out?.restore.isEmpty == true)
    }
}
